### PR TITLE
merge feature test usart into release

### PR DESCRIPTION
**Pull request by PatrickDekker98**

### CODE REVIEW

File: code/headers/hardware_usart.hpp [hunk 1]
```diff
@@ -27,18 +28,7 @@ namespace r2d2 {
...
-        ///
-        /// This is the function used for console (ostream) output.
-        /// The embedded targets provide an implementation that writes
-        /// to the serial port.
-        ///
-        /// This definition is weak, which allows
-        /// an application to provide its own definition.
-        void uart_putc(char c);
-
+        hardware_usart_c(unsigned int baudrate, uart_ports_c &usart_port) ;
```
Comment (itzandroidtab) [inline code/headers/hardware_usart.hpp]: why is this by reference and the variable is not.

File: code/headers/test_usart.cpp [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    bool test_usart_c::send(const uint8_t c) {
+        //Not needed in test implementation
+    }
+
+    void test_usart_c::putc(char c) {
+        send(c);
+    }
+
+    uint8_t test_usart_c::receive() {
+        return 0xAA;
```
Comment (itzandroidtab) [inline code/headers/test_usart.cpp:27]: Why does this return 0xAA can you please explain why 0xAA?

File: code/headers/test_usart.cpp [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    bool test_usart_c::char_available() {
+        return true;
+    }
+
+    char test_usart_c::getc() {
+        return receive();
+    }
+
+    unsigned int test_usart_c::available() {
+        return 1;
```
Comment (itzandroidtab) [inline code/headers/test_usart.cpp:39]: why is this a unsigned int? not a size_t if it returns the amount of data available and bool if it only returns if data is available.

File: code/headers/test_usart.hpp [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    class test_usart_c : public usart_connection_c {
+    private:
+        unsigned int boudrate;
+        uart_ports_c &usart_port;
+        queue_c<uint8_t, 250> input_buffer;
+
+
+    public:
+
+        test_usart_c(unsigned int baudrate, uart_ports_c controller)
```
Comment (itzandroidtab) [inline code/headers/test_usart.hpp]: why is this not a reference? the variable is a reference in the class

File: code/headers/usart_connection.hpp [hunk 5]
```diff
@@ -0,0 +1,53 @@
+/// @file
+/// @brief      base class for a usart implementation hardware, software and mock
```
Comment (itzandroidtab) [inline code/headers/usart_connection.hpp:2]: is this doxygen for the file or the class?

File: code/main.cpp [hunk 6]
```diff
@@ -9,29 +9,5 @@ int main(void) {
...
-                complete_string = false;
-            } else {
-                n++;
-                gps_buffer[n]  =cc;
-            }
-        }
-
-    }
-
+    r2d2::hardware_usart_c c = r2d2::hardware_usart_c(9600, port);
```
Comment (itzandroidtab) [inline code/main.cpp]: can you add comments what everything does in here?

File: code/headers/hardware_usart.hpp [hunk 7]
```diff
@@ -1,17 +1,18 @@
 #pragma once
 
-#include "hwlib.hpp"
+#include <hwlib.hpp>
 #include <queue.hpp>
 #include <uart_ports.hpp>
+#include <usart_connection.hpp>
```
Comment (itzandroidtab) [inline code/headers/hardware_usart.hpp:6]: why are you including a file that is in the local directory as a library?

File: code/headers/hardware_usart.hpp [hunk 8]
```diff
@@ -1,17 +1,18 @@
 #pragma once
 
-#include "hwlib.hpp"
+#include <hwlib.hpp>
 #include <queue.hpp>
 #include <uart_ports.hpp>
```
Comment (itzandroidtab) [inline code/headers/hardware_usart.hpp:5]: this one as well

**Review by itzandroidtab — CHANGES_REQUESTED** [review]

File: code/headers/test_usart.cpp (hunk 3 above)
Comment (PatrickDekker98) [reply code/headers/test_usart.cpp:39]: because the base class returns a unsigned int

File: code/headers/test_usart.hpp (hunk 4 above)
Comment (PatrickDekker98) [reply code/headers/test_usart.hpp]: will fix

File: code/headers/hardware_usart.hpp (hunk 8 above)
Comment (PatrickDekker98) [reply code/headers/hardware_usart.hpp:5]: check the style guide

File: code/headers/hardware_usart.hpp (hunk 7 above)
Comment (PatrickDekker98) [reply code/headers/hardware_usart.hpp:6]: What do you mean? according to the style guide we need to iclude with <> not ""

File: code/headers/test_usart.cpp (hunk 2 above)
Comment (PatrickDekker98) [reply code/headers/test_usart.cpp:27]: Because a byte was needed and this is a byte

File: code/headers/hardware_usart.hpp (hunk 7 above)
Comment (gerritvanos) [reply code/headers/hardware_usart.hpp:6]: in the build environment we include everything with <> in theory both will compile perfectly fine because the compiler findes the file both ways

File: Makefile.inc [hunk 9]
```diff
@@ -19,4 +19,4 @@ HEADERS += $(wildcard $(USART)/code/headers/*.hpp)
 
 # search for queue.hpp
 SEARCH  += $(USART)/../datastructures/code/headers
-HEADERS += $(wildcard $(USART)/../datastructures/code/headers/queue.hpp)
\ No newline at end of file
+HEADERS += $(wildcard $(USART)/../datastructures/code/headers/queue.hpp)
```
Comment (gerritvanos) [inline Makefile.inc:22]: this is probably not necessary to include datastructures since it will be included by default from internal_communication but to be complete it is nice to have.

File: code/headers/test_usart.cpp [hunk 10]
```diff
@@ -0,0 +1,41 @@
+#include "test_usart.hpp"
```
Comment (gerritvanos) [inline code/headers/test_usart.cpp]: the .cpp file is located in the headers folder?

File: code/headers/test_usart.cpp [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+
+namespace r2d2 {
+
+    test_usart_c::test_usart_c(unsigned int baudrate, usart_ports_c controller):
+        baudrate(baudrate),
+        controller(controller)
+    {}
+
+    void test_usart_c::enable(){
```
Comment (gerritvanos) [inline code/headers/test_usart.cpp:10]: if you define the functions in the base class without the =0 you dont need to overide the functions with empty body not necessary to change for now but might be usefull in the future

File: code/headers/hardware_usart.hpp (hunk 1 above)
Comment (PatrickDekker98) [reply code/headers/hardware_usart.hpp]: fixed

File: Makefile.inc (hunk 9 above)
Comment (PatrickDekker98) [reply Makefile.inc:22]: I believe it does not pass travis without this

File: code/headers/test_usart.cpp (hunk 10 above)
Comment (PatrickDekker98) [reply code/headers/test_usart.cpp]: automated clion stuff sorry about that

File: code/headers/test_usart.cpp (hunk 11 above)
Comment (PatrickDekker98) [reply code/headers/test_usart.cpp:10]: welp, I removed these overrides because I got an undefined vtable error, will add them

File: code/main.cpp (hunk 6 above)
Comment (PatrickDekker98) [reply code/main.cpp]: done

File: Makefile.inc (hunk 9 above)
Comment (gerritvanos) [reply Makefile.inc:22]: it theory this does not influence travis since the inc is only used when including as a library but as said voor (volledigheid) is het wel netjes om te hebben.

File: code/headers/test_usart.cpp (hunk 11 above)
Comment (gerritvanos) [reply code/headers/test_usart.cpp:10]: this is better but did not mean this i was not very clear sorry, i meant that if you dont make them =0 in the base class you can basicly skip this in de inherited classes if you dont need the functions but this solves my problem for now.

File: code/headers/usart_connection.hpp (hunk 5 above)
Comment (PatrickDekker98) [reply code/headers/usart_connection.hpp:2]: file

File: code/headers/hardware_usart.hpp [hunk 12]
```diff
@@ -51,37 +41,37 @@ namespace r2d2 {
         /// used for char arrays to quickly send more than one byte
         /// In practice useing this fuction is more stable than repeated use of
         /// send_byte() Especially when the values are repeated.
-        hardware_usart_c &operator<<(const char *c);
+        hardware_usart_c &operator<<(const char *c) ;
```
Comment (itzandroidtab) [inline code/headers/hardware_usart.hpp:44]: Why did you add a space after the brackets?

File: code/src/hardware_usart.cpp [hunk 13]
```diff
@@ -50,6 +51,7 @@ namespace r2d2 {
 
             PMC->PMC_PCER0 = (0x01 << ID_USART3);
         } else if (usart_port == uart_ports_c::uart0) {
+
```
Comment (itzandroidtab) [inline code/src/hardware_usart.cpp:54]: Maybe something like a HWLIB_PANIC_WITH_LOCATION here

**Comment by itzandroidtab** [comment]
It looks okay. I just don't like that you use char and unsigned int together with uint8_t. You should decide which of the style you are gonna use. I am gonna approve of this for now but i kinda expect you to change to one of the two styles.

**Review by itzandroidtab — APPROVED** [review]
